File: Beginner/SearchInsertPosition/SearchInsertPosition.py
from typing import List

# O(logn) solution using binary search (already sorted so it's not NlogN)
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums)-1

        while l <= r:
            mid = (l+r)//2
            if nums[mid] == target:
                return mid
            if target > nums[mid]:
                l = mid+1
            if target < nums[mid]:
                r = mid-1
        return l

s = Solution()
print(s.searchInsert([1,3,5,6], 5))
print(s.searchInsert([1,3,5,6], 2))
print(s.searchInsert([1,3,5,6], 7))

File: Beginner/SearchInsertPosition/SearchInsertPosition.cpp
#include <vector>
#include <iostream>

class Solution {
    public:
        int searchInsert(std::vector<int>& nums, int target) {
            int left = 0;
            int right = nums.size()-1;
            
            while (left <= right) {
                int mid = (left + right) / 2;
                if (target == mid) {
                    return mid;
                } else if (target > nums[mid]) {
                    left = mid + 1;
                } else {
                    right = mid-1;
                }
            }
            return left;
        }
};

int main() {
    Solution s;
    std::vector<int> nums = {1, 3, 5, 6};
    int target = 5;
    std::cout << s.searchInsert(nums, target) << std::endl;
    return 0;
}